### ISSUE TITLE
Heterogeneous systems setting

### ISSUE THREAD

**Issue by ghost** [epfml/federated-learning-public-code#1]
Hi,

First of all, excellent work! But I have some questions about Figure 4 in Section 4.3. Hope you can answer them for me.

What does the ensembled model refer to? What kind of architecture did you use for the ensembled model in each experiment? If I guess right, you applied ensemble distillation to each prototype with averaged logits from models of all prototypes with the result of one global model for each prototype. What is the exact ensemble method that you used to test these global models over the test dataset? Averaging the logits and then activate the output? Or something similar to the vote strategy? I don't think you point this out in the paper.

**Comment by tlin-taolin** [comment]
Hi, sorry for the late reply (I did not receive the email notification and I came across this issue when I was checking this repository today).

Yes, as mentioned in Algorithm 3 (Appendix), we applied ensemble distillation to each prototype, where the averaged logits from *models of all prototypes* were used for ensemble distillation on each prototype.

Indeed, I did not understand your question about "What is the exact ensemble method that you used to test these global models over the test dataset?". Once we retrieved the distilled model for each prototype (we will then drop the received models), we individually tested these distilled prototype models on the test dataset. 
It is the same as the homogeneous case: homogeneous case uses the averaged logits from models of one prototype, while heterogeneous uses the averaged logits from models of all prototypes. 
It is not related to "Averaging the logits and then activate the output? Or something similar to the vote strategy?"

Hope it helps.

**Comment by ghost** [comment]
Sorry, I misunderstood some parts back then. But still, what is the exact prototype you choose for the red lines in Figure 4?

**Comment by tlin-taolin** [comment]
The `Ensembled model` refers to an impractical performance upper bound case where we will keep the received models (different prototypes) and perform softmax over the averaged logits (no matter the prototype) from these models. We aim to use to better understand the performance drop caused by the distillation via an unlabeled dataset.

We will clarify this point later (thanks for pointing out).

**Comment by ghost** [comment]
Oh, thanks a lot. Now I see it more clearly. I recently followed your work by replacing the unlabeled dataset with data-free techniques. But it does not have such a performance promotion as your experiments. I think the prior knowledge in the unlabeled dataset may take most of the credits for that. Do you think that maybe in some real-life cases, the proper unlabeled datasets cannot always be found? What is your strategy in choosing them?
